xarray/core/dataset.py

```diff
@@ -3870,8 +3870,8 @@ def stack(
     def to_stacked_array(
         self,
         new_dim: Hashable,
-        sample_dims: Sequence[Hashable],
-        variable_dim: str = "variable",
+        sample_dims: Mapping[Hashable, int],
```

Comment:
Is this correct, though? The docstring and the example below (https://github.com/pydata/xarray/pull/5494/files#diff-763e3002fd954d544b05858d8d138b828b66b6a2a0ae3cd58d2040a652f14638R3932) has it as a sequence?

Reply:
I really don't know. mypy doesn't complain with this at least on my specific issue.

Seems to be several versions what these dim arguments can be:
https://github.com/pydata/xarray/blob/382fe69750988f9cf54b4ac7aa0f0ddc4b871835/xarray/core/dataarray.py#L1851

Reply:
This function starts with `stacking_dims = tuple(dim for dim in self.dims if dim not in sample_dims)` — is there a working case of this function taking a differently argument?

I can empathize that mypy doesn't fail, but that can't be the only constraint — imagine there's already a typing mistake in the library — it might be possible to make mypy pass by adding an additional mistake.

Reply:
I think I get it now. This is the code that causes this error:
```python
arr = arr.to_stacked_array(
    variable_dim=dim, new_dim="stacked__dim", sample_dims=arr.dims
).unstack("stacked__dim")
```
arr is a Dataset in this case.

DataArrays return `a.dims` as a tuple:
```python
a = xr.DataArray(
    [[0, 1, 2], [3, 4, 5]],
    dims=("dim_0", "dim_1"),
    coords=dict(dim_0=(["dim_0"], [0, 1]), dim_1=(["dim_1"], ["u", "v", "w"])),
)
a.dims
Out[16]: ('dim_0', 'dim_1')
```

But Datasets return ds.dims as a sorted dict:
```python
ds = xr.Dataset({0: ("dim_0", [1]), 1: ("dim_0", [2]), 2: ("dim_0", [3])})
Out[14]: 
<xarray.Dataset>
Dimensions:  (dim_0: 1)
Dimensions without coordinates: dim_0
Data variables:
    0        (dim_0) int32 1
    1        (dim_0) int32 2
    2        (dim_0) int32 3

ds.dims
Out[12]: Frozen(SortedKeysDict({'dim_0': 1}))

# ds.dims is a valid input type:
ds.to_stacked_array(
    variable_dim="cartesian", new_dim="variable", sample_dims=ds.dims
)
<xarray.DataArray 0 (dim_0: 1, variable: 3)>
array([[1, 2, 3]])
Coordinates:
  * variable   (variable) MultiIndex
  - cartesian  (variable) int64 0 1 2
Dimensions without coordinates: dim_0
```

Reply:
Yes great! I should have been clearer on the other PR — I had tried to indicate that this might be the reason.

Does `Collection` work? Conceptually it's not so much a union of sequence + mapping as "something we can iterate over to get dim names, and test for containing items".

Reply:
Seems so:

```python
if isinstance(ds.dims, Collection):
    print("hey")
hey

if isinstance(a.dims, Collection):
    print("hey")
hey
```

But `Collection` is barely used anywhere in the xarray code, so it would be a style change for a rather popular argument.

Reply:
Thanks!

I _think_ with these:
- If passing in a dict is OK — i.e. the functions that are applied only see the keys — then we should use `Collection`. That's conceptually simpler than `Sequence` or `Mapping` — it's the properties of `Collection` that are required, rather than listing its subclasses.
- If it's not OK, we should use `Sequence`

That said, I often need to remind myself of how python treats these, and so I'm not that confident. Feel free to do what you think is best, and maybe add a link to this discussion in the code for posterity.

Reply:
I've changed it to Collection now. 
But I think a lot of these dim-arguments in other functions could be changed to Collection as well. Right now this change is the odd one out. 
But that's for a future PR (not by me).